server/service/schedule/schedule.go

```diff
@@ -0,0 +1,219 @@
+package schedule
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/getsentry/sentry-go"
+	kitlog "github.com/go-kit/kit/log"
+	"github.com/go-kit/log/level"
+)
+
+type schedule struct {
+	ctx        context.Context
+	name       string
+	instanceID string
+	Logger     kitlog.Logger
+
+	muChecks       sync.Mutex
+	configInterval time.Duration
+	schedInterval  time.Duration
+	locker         Locker
+	configCheck    func(start time.Time, wait time.Duration) (*time.Duration, error)
+	preflightCheck func() bool
+
+	muJobs sync.Mutex
+	jobs   map[string]job
+}
+
+func (s *schedule) getConfigInterval() time.Duration {
+	s.muChecks.Lock()
+	defer s.muChecks.Unlock()
+	return s.configInterval
+}
+
+func (s *schedule) setConfigInterval(interval time.Duration) {
+	s.muChecks.Lock()
+	defer s.muChecks.Unlock()
+	s.configInterval = interval
+}
+
+func (s *schedule) getSchedInterval() time.Duration {
+	s.muChecks.Lock()
+	defer s.muChecks.Unlock()
+	return s.schedInterval
+}
+
+func (s *schedule) setSchedInterval(interval time.Duration) {
+	s.muChecks.Lock()
+	defer s.muChecks.Unlock()
+	s.schedInterval = interval
+}
+
+func (s *schedule) SetPreflightCheck(fn func() bool) {
+	s.muChecks.Lock()
+	defer s.muChecks.Unlock()
+	s.preflightCheck = fn
+}
+
+func (s *schedule) SetConfigCheck(fn func(start time.Time, wait time.Duration) (*time.Duration, error)) {
```

Comment:
The naming is a little confusing, because it doesn't communicate when/why the fn function gets called. It looks like it should return the new interval for the job. In order to do so, it needs to look up the latest AppConfig. Perhaps `fn` should just take `AppConfig` directly, instead of having to look it up. Also, it seems like `*schedule.run` already handles the start and previous interval logic.

I think that the function signature should look more like

```
// The fn gets called when AppConfig has changed and should return a new interval.
func (s *schedule) GetInterval(fn func(cfg fleet.AppConfig) (time.Duration, error)) {
```

The duration returned doesn't need to be a pointer. You can return 0 in case of an error.